homeassistant/components/drop_connect/coordinator.py

```diff
@@ -0,0 +1,183 @@
+"""DROP device data update coordinator object."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.util.json import JSON_DECODE_EXCEPTIONS, json_loads
+
+from .const import (
+    CONF_DATA_TOPIC,
+    CONF_DEVICE_DESC,
+    CONF_DEVICE_ID,
+    CONF_DEVICE_TYPE,
+    CONF_HUB_ID,
+    DEV_HUB,
+    DOMAIN,
+    KEY_STATUS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class DROPDeviceDataUpdateCoordinator(DataUpdateCoordinator):
+    """DROP device object."""
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+        """Initialize the device."""
+        self.config_entry: ConfigEntry = config_entry
+        self._model: str = f"{config_entry.data[CONF_DEVICE_DESC]} on hub {config_entry.data[CONF_HUB_ID]}"
+        if config_entry.data[CONF_DEVICE_TYPE] == DEV_HUB:
+            self._model = f"Hub {config_entry.data[CONF_HUB_ID]}"
+        self._manufacturer: str = "Chandler Systems, Inc."
+        self._device_name: str = config_entry.data["name"]
+        self._device_information: dict[str, Any] = {}
+        super().__init__(hass, _LOGGER, name=f"{DOMAIN}-{config_entry.unique_id}")
+
+    async def DROPMessageReceived(
+        self, topic: str, payload: str, qos: int, retain: bool
+    ) -> None:
+        """Process a received MQTT message."""
+        topicRoot = self.config_entry.data[CONF_DATA_TOPIC].removesuffix("/#")
+        try:
+            jsonData = json_loads(payload)
+            if topic.startswith(topicRoot):
+                structureKey = topic.removeprefix(topicRoot).removeprefix("/")
+                _LOGGER.debug(
+                    "New data for %s/%s [%s]: %s",
+                    self.config_entry.data[CONF_HUB_ID],
+                    self.config_entry.data[CONF_DEVICE_ID],
+                    structureKey,
+                    payload,
+                )
+
+                # Create empty dictionary for this structure key if it does not already exist.
+                if structureKey not in self._device_information:
+                    self._device_information[structureKey] = {}
+
+                # Merge incoming data into the existing dictionary.
+                self._device_information[structureKey].update(jsonData)
+                self.async_set_updated_data(None)
```

Comment:
Even this may work, it is more common to implement a `_async_update_data()`, which cares about updated data, stores them into `coordinator.data` property (_no need of all these additional properties down here_) and signals to all registered entities that updated data are available.
The [proximity coordinator](https://github.com/home-assistant/core/blob/dev/homeassistant/components/proximity/coordinator.py) is a good example of how to use the `_async_update_data()` with external triggers instead of scheduled updates.

Reply:
I've been studying the integration you cited and so far haven't worked out how I would trigger an update when a new MQTT message is received. As it is, the DROP hub publishes on the MQTT topic only when data has changed, so the `drop_message_received()` function is likewise only called when new data is pushed. On the surface, it seems safe and efficient to simply update the stored data and notify HA that something has changed.

Also, I'd prefer to intentionally keep a layer between the API keys and the data properties in the coordinator to maintain brevity in the API and readability in the code. (e.g. use `coordinator.average_water_used` within the integration, but it pulls from 'avgUsed' in the API data)  There is the potential for some really cryptic API keys as we add more functionality from our products into the integration.

So, I guess my question is whether or not the approach I'm currently using is acceptable, even though there might be a more elegant solution...

Reply:
this can also be done with an `dataclass` where all properties could be collected (_and also be typed_), filled with data and put into `coordinator.data` by an `_async_update_data()` or `async_set_updated_data(data)`. Sure static device info like model, manufacturer and name can still be static defined properties.
But this could also be a topic for a next improvement PR 😉

Reply:
I've put this on my to-do list, and there will definitely be future improvement PRs.